test_proxy/client_handler.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the client handler process for proxy_server.py.
+"""
+import os
+
+from google.cloud.environment_vars import BIGTABLE_EMULATOR
+from google.cloud.bigtable.data import BigtableDataClientAsync
+
+
+def error_safe(func):
+    """
+    Catch and pass errors back to the grpc_server_process
+    Also check if client is closed before processing requests
+    """
+    async def wrapper(self, *args, **kwargs):
+        try:
+            if self.closed:
+                raise RuntimeError("client is closed")
+            return await func(self, *args, **kwargs)
+        except (Exception, NotImplementedError) as e:
+            # exceptions should be raised in grpc_server_process
+            return encode_exception(e)
+
+    return wrapper
+
+def encode_exception(exc):
+    """
+    Encode an exception or chain of exceptions to pass back to grpc_handler
+    """
+    from google.api_core.exceptions import GoogleAPICallError
+    error_msg = f"{type(exc).__name__}: {exc}"
+    result = {"error": error_msg}
+    if exc.__cause__:
+        result["cause"] = encode_exception(exc.__cause__)
+    if hasattr(exc, "exceptions"):
+        result["subexceptions"] = [encode_exception(e) for e in exc.exceptions]
+    if hasattr(exc, "index"):
+        result["index"] = exc.index
+    if isinstance(exc, GoogleAPICallError):
+        if exc.grpc_status_code is not None:
+            result["code"] = exc.grpc_status_code.value[0]
+        elif exc.code is not None:
+            result["code"] = int(exc.code)
+        else:
+            result["code"] = -1
+    elif result.get("cause", {}).get("code", None):
+        # look for code code in cause
+        result["code"] = result["cause"]["code"]
+    elif result.get("subexceptions", None):
+        # look for code in subexceptions
+        for subexc in result["subexceptions"]:
+            if subexc.get("code", None):
+                result["code"] = subexc["code"]
```

Comment:
Yes, we make use of [exception groups](https://peps.python.org/pep-0654/) so that we can expose a set of retried errors together with the final error. The exceptions are wrapped in a `bigtable.RetryException`. The final exception in the group should be the most relevant one

More details here: https://github.com/googleapis/python-bigtable/pull/825